### PR TITLE
Chain Service health checks

### LINKED ISSUE #1202: Create health check for ChainService
As part of #1183, all services should maintain a Status method that will return an error if there is any unhealthy status to report. This task requires some creative thinking to evaluate a definition of a healthy service. It is possible that a service can always be healthy and in that case we would simply remove the TODO and update the godoc to explain why the service is always healthy.

Service definition: beacon-chain/blockchain/service.go

What makes this service healthy or unhealthy?

Some ideas:
- The web3service has gone away and we are trying to reconnect to it.
- Chain head block not received in some time
- Block processing hasn't received a block in some time

### PR DESCRIPTION
**Pull request by ratikesh9**
Resolve #1202 

---

# Description

Adding Web3service, Bolt db, channel feed specific status check

### CODE REVIEW

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -77,8 +78,16 @@ func (c *ChainService) Start() {
 	c.genesisTime, err = c.beaconDB.GenesisTime()
 	if err != nil {
 		log.Fatalf("Unable to retrieve genesis time - blockchain service could not start: %v", err)
+		c.failStatus = err
 		return
 	}
+
+	if c.web3Service.Status() != nil {
+		log.Fatalf("Unable to connect to web3Service")
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/service.go]: Fatal will crash the server.

File: beacon-chain/blockchain/service.go [hunk 2]
```diff
@@ -77,8 +78,16 @@ func (c *ChainService) Start() {
...
+		c.failStatus = c.web3Service.Status()
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/service.go]: You don’t need to set failStatus for something you can query within the status func.

Your Status() should look like 

```go
func (c * ChainService) Status() error {
  if c.web3Service.Status() != nil {
    return c.web3Service.Status()
  } 

  ... other checks here ...
 
  return nil
}
```

File: beacon-chain/blockchain/service.go [hunk 3]
```diff
@@ -140,6 +153,7 @@ func (c *ChainService) updateHead(processedBlock <-chan *pb.BeaconBlock) {
 			return
 		case block := <-processedBlock:
 			if block == nil {
+				c.failStatus = fmt.Errorf("chain head block not receieved for updating ")
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/service.go]: I don’t think this is necessarily an unhealthy condition. An empty thing sent through the channel can be ignored.

Also, is this failure permenant? What about when the service recovers?

File: beacon-chain/blockchain/service.go [hunk 4]
```diff
@@ -228,6 +242,10 @@ func (c *ChainService) blockProcessing(processedBlock chan<- *pb.BeaconBlock) {
 		// can be received either from the sync service, the RPC service,
 		// or via p2p.
 		case block := <-c.incomingBlockChan:
+			if block == nil{
+				c.failStatus = fmt.Errorf("block not received for processing ")
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/service.go]: Same comment here

File: beacon-chain/blockchain/service.go (hunk 2 above)
Comment (ratikesh9) [reply beacon-chain/blockchain/service.go]: Will correct this

File: beacon-chain/blockchain/service.go (hunk 3 above)
Comment (ratikesh9) [reply beacon-chain/blockchain/service.go]: Understood something wrong. Agreed it will impact normal functioning. Will check these

**Comment by nisdas** [comment]
@ratikesh9 Hey what's the status on this ?

**Comment by ratikesh9** [comment]
Will push an update on this by today.

**Comment by prestonvanloon** [comment]
Hey @ramisra, what happened here? Do you plan to continue this feature?

**Pull request closed without merging**